core/db/utils.py

```diff
@@ -0,0 +1,44 @@
+import time
+from functools import wraps
+from typing import Type
+
+from flask import current_app
+
+from core.db import db
+
+
+def database_transaction(max_retries: int, sleep_duration: float, error_type: Type[Exception]):
+    """Execute a transaction with retries for a specified error type.
+
+    This function wraps the provided 'function' in a database transaction and allows for retries
+    in case of a specified error type. If an error of the specified type occurs during the
+    transaction, the operation is retried up to a specified maximum number of times with a
+    delay between retries.
+
+    :param max_retries: The maximum number of retries in case of an error.
+    :param sleep_duration: The duration to sleep between retry attempts.
+    :param error_type: The type of error to catch and retry.
+    :return: None
+    """
+
+    def decorator(func):
+        @wraps(func)
+        def wrapper(retries_left=max_retries, *args, **kwargs):
```

Comment:
If it's in a transaction what's the benefit of re-trying, are the transactions not atomic?

Reply:
Because I don't think it will queue by default. When I've tested locally it falls over if you have a database transaction open and another user tries to enter the same loop. 

We want to retry based on the approximate duration of the function call, and a reasonable number attempts (e.g. 5 if we expect a maximum of 5 users to be concurrently submitting returns).